streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,215 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
     List<StreamTask> allTasks() {
         final List<StreamTask> tasks = super.allTasks();
         tasks.addAll(restoring.values());
+        tasks.addAll(suspended.values());
         return tasks;
     }
 
     @Override
     Set<TaskId> allAssignedTaskIds() {
         final Set<TaskId> taskIds = super.allAssignedTaskIds();
         taskIds.addAll(restoring.keySet());
+        taskIds.addAll(suspended.keySet());
         return taskIds;
     }
 
     @Override
     boolean allTasksRunning() {
-        return super.allTasksRunning() && restoring.isEmpty();
+        return super.allTasksRunning() && restoring.isEmpty() && suspended.isEmpty();
     }
 
-    RuntimeException closeAllRestoringTasks() {
+    @Override
+    void closeTask(final StreamTask task, final boolean clean) {
+        if (suspended.containsKey(task.id())) {
+            task.closeSuspended(clean, false, null);
+        } else {
+            task.close(clean, false);
+        }
+    }
+    
+    Set<TaskId> suspendedTaskIds() {
+        return suspended.keySet();
+    }
+
+    RuntimeException suspend(final Set<TaskId> revokedTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+
+        log.trace("Suspending running {} {}", taskTypeName, runningTaskIds());
+        firstException.compareAndSet(null, suspendTasks(revokedTasks, revokedChangelogs));
+
+        log.trace("Close created {} {}", taskTypeName, created.keySet());
+        firstException.compareAndSet(null, closeNonRunningTasks(revokedTasks, revokedChangelogs));
+
+        return firstException.get();
+    }
+
+    private RuntimeException suspendTasks(final Set<TaskId> tasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        for (final Iterator<TaskId> it = tasks.iterator(); it.hasNext(); ) {
+            final TaskId id = it.next();
+            final StreamTask task = running.get(id);
+
+            if (suspended.containsKey(id)) {
+                // if the task is already suspended just remove it and move on
+                it.remove();
```

Comment:
I'm having some trouble following the workflow here and elsewhere. We're passing in the set of tasks, and then removing elements from them while we're working on them. But then we continue to use the same collection elsewhere.

It's not clear what, if anything, would be left in the collection by the time L92, for example is invoked, or even more mysterious, TaskManagerL266 also uses this collection later on.

It's a bit complicated because normally, we'd return the "resulting" collection instead of mutating a parameter, but we're already returning the "first exception". I'm not sure the best way to resolve this...

Reply:
Note there may not be a great solution, either, just wanted to let you know I found it confusing.

Reply:
Basically we have a set of tasks that are revoked, and we need to suspend or close them differently based on whether they're in suspended, restoring, created, etc. We should verify at the end that all revoked tasks have been dealt with. So this task set is really the set of tasks that still need to be closed/suspended/revoked. I'll try to think up a less confusing solution (do you have any ideas?) but for starters what if we renamed it to something like `tasksToRevoke` ? 
I definitely do agree it's broken up kind of awkwardly at the moment, and can be cleaned up given this refactoring eg so that `suspend` only suspends.

Reply:
I see. Maybe just the name could eliminate the confusion, like `remainingRevokedTasksToHandle`, then it seems more natural that it should empty by attrition in subsequent method calls?

I didn't have a good idea before, but I've just had one thought... Can we partition the set based on those states up front, so then we'd get a set of suspended tasks to handle, a disjoint set of restoring ones, etc., and each method can just deal with its own set?

Reply:
Yeah I rejected that at first because I couldn't think of a clean way to make sure we didn't miss any. But, obviously we can just figure that out when we partition them...

Reply:
Cleaned up this part, hopefully it's less confusing now